odoo/tools/sql.py

```diff
@@ -23,23 +23,108 @@
     'SET DEFAULT': 'd',
 }
 
+
+class SQL:
+    """ An object that wraps an SQL string with its positional parameters.
+    The positional parameters can be either actual parameters or SQL objects
+    themselves. In the latter case, those SQL objects are merged into the main
+    SQL object, and their respective parameters are merged at the right position.
+    """
+    __slots__ = ['__string', '__args']
+
+    def __init__(self, string: str = "", *args):
+        if isinstance(string, SQL):
+            string, args = string.__string, string.__args
+        string % tuple("x" for arg in args)
+        self.__string = string
+        self.__args = args
+
+    @property
+    def string(self) -> str:
+        """ Return the final SQL code string. """
+        return self.__string % tuple(
+            arg.string if isinstance(arg, SQL) else "%s"
+            for arg in self.__args
+        )
```

Comment:
Thanks for the suggestion, I will have a look at this one.

Reply:
Here is my take on this suggestion. We currently have the following equivalence:
```py
cr.execute(sql)                     # implicit: the purpose of SQL
cr.execute(sql.string, sql.params)  # explicit: straightforward mapping
```
I don't see much the added value of having method `as_string()`.  The purpose of `as_string()` is to "compile" a `Composable` into a `str`, but this only concerns the code part of the query, not the parameters.  If we follow that convention, it would add the following weird equivalences to the ones above:
```py
cr.execute(sql.as_string(cr.connection), sql.params)  # explicit: confusing
cr.execute(sql, sql.params)                           # implicit: even more confusing
```
If instead, we make `as_string()` behave as `cr.mogrify()`, it would add:
```py
cr.execute(sql.as_string(cr.connection))  # explicit: feels weird without parameters
cr.execute(sql)                           # implicit: could replace the hook in execute()
```
So, what's the point?